WIP: PodcastIndex transcription

tonytamsf opened the pull request:
Fix https://github.com/AntennaPod/AntennaPod/issues/4935

This is a work in progress, do not review yet.  I'm using this PR to discuss questions that are coming up when I'm implementing

### As of Oct 25th
- store transcript in database when downloaded

### As of Oct 18th
- I'm able to display the transcript in a WebView and highlight the transcript along with the audio (a little choppy right now)
  (this is just a test for myself on how Webview and HTML anchors will work to show the words). See demo and screenshot https://github.com/AntennaPod/AntennaPod/pull/6710#issuecomment-1769744558

### As of Oct 15th
- I'm able to fetch the SRT and JSON format and parse the content
- Display raw data at the end of shownotes

### TODO
~~- Store/Fetch the transcript in the database~~
- wait for UI design and incorporate the transcript

### Podcasts sampling
- Roblox Tech Talk - `text/plain` https://share.transistor.fm/s/39ce510f/transcript.txt
   -  parsable with timecoees
-  Dev Interrupted - `text/html` https://feeds.buzzsprout.com/1422892/10190284/transcript
   - not parsable
- The Method - json & 'text/html' https://feeds.buzzsprout.com/1859927/9638130/transcript
   - html not parsable
   - json is fine https://feeds.buzzsprout.com/1859927/9638130/transcript.json
   - (only 2 episode)
- Radical Candor - `application/x-subrip` https://freepodcasttranscription.com/transcription/5e6070b67882e784d6b5baf4b19976bb717f1115.srt
   - parsable
- Indie Bites - `text/plain` https://share.transistor.fm/s/6cdce2f8/transcript.txt (no timecodes)
  - not parsable
- The Golden Hour - `text/html` https://www.podcastone.com/episode/Someones-Getting-Tased--The-Golden-Hour-51-w/-Brendan-Schaub-Erik-Griffin--Chris-DElia?view=transcript
   - not parsable

ByteHamster commented:
> Store the transcript in the database

The transcripts are quite large. I would do what we already do for `podcast:chapters`: just load them when displaying the transcripts page

tonytamsf commented:
> The transcripts are quite large. I would do what we already do for podcast:chapters: just load them when displaying the transcripts page

Got it, thanks for the suggestion.  I agree

keunes commented:
> what we already do for podcast:chapters: just load them when displaying the transcripts page

The original idea was to store it (in the database) while the episode is downloaded, and purge it when the media file gets deleted. Not sure why we stepped away from that approach.

In any case, I would be a strong advocate for taking this approach (also) for transcripts. We cannot work based on the assumption that there will always be internet to download this 'on the fly'. There are still large/many areas in the world where this is not the case and this also when travelling (e.g. without roam like at home or in an airplane) this won't work.

If you download an episode, I think, we should download the whole thing. If that's too much for your device, you'll probably stream anyway (b/c the media files are much heavier).

tonytamsf commented:
@keunes and @ByteHamster I don't have a strong opinion here about the storage of the transcript. I am a little hesitant to store large blobs of data in SQLite, but the docs for SQLite claims there are no issues

https://www.sqlite.org/limits.html

ByteHamster commented:
What we did for the chapters is to cache the file while downloading. I intentionally didn't mention this earlier because it is not needed for a first implementation. As a first step, just loading the transcript when opening the page is easier and already has the right code structure to do the caching later. I'm sure there is a lot more to discuss in this PR, so I wanted to keep it simple for now.

keunes commented:
I don't mind how it's stored (database or cache or whatever). Or whether this would be done in a separate PR. But I feel really quite strongly about releasing this in production without the approach I described (media file download = meta data download).

Maybe we can make a `transcripts` branch and make it the target for this and other PRs, if that makes it easier to review through separate PRs. Could be, like, three: this (fetch), store & UI. Then we can merge from the `transcripts` branch to `development`.

tonytamsf commented:
While waiting for a mock up, I went ahead with an implementation in order to test out using HTML anchors to be jump to the word/phrase being side in audio. cc @ByteHamster and @keunes 

Here is a quick video demo

https://github.com/AntennaPod/AntennaPod/assets/149837/d338649a-42ee-44d4-a8dd-5631dec6a503


1) Here is a chrome debug view of the words in transcript stored as `<a id="seg####"`

<img width="1662" alt="Screenshot 2023-10-18 at 6 15 40 PM" src="https://github.com/AntennaPod/AntennaPod/assets/149837/44b9fcad-7fbd-4a19-b02a-0ac65bf364ca">

2) I added a 'Transcript' button on the main playing screen (obvious that doesn't really work in spacing)

![screenshot-first](https://github.com/AntennaPod/AntennaPod/assets/149837/bf9f3227-57b7-42ab-ad3d-af139cd2a993)

3) I added a 3rd page (after shownotes) where I show the transcript using a WebView.  You can see the 'Yellow' is the highlighted word.  I need to figure out how to make the visuals work when the host is speaking multiple words a second since we refresh the screen only once a second.

![screenshot-sedond](https://github.com/AntennaPod/AntennaPod/assets/149837/935b4226-a9d7-423d-9b19-6a2fa1818fcb)

ByteHamster commented:
Hmm, I'm not really sure about using a WebView for this. I think we rather want to use TextViews in a RecyclerView. Refreshing a WebView every second will probably cause us trouble later (eg when users scroll while it is refreshing).

Later, we should probably also add things like scrolling the screen automatically while playing, so that the current section always stays visible. Additionally, we should detect when the user deliberately scrolls away from the current section and then stop scrolling automatically. These things are probably out of scope for a first implementation (this PR will be quite large already with just the basic features), but we should already keep them in mind. I think a WebView makes these things quite hard to implement.

tonytamsf commented:
@ByteHamster I will try to implement RecyclerView+TextView and see how that works. The WebView worked okay with anchor tags, but I agree that refreshing every second could be a problem.  It was pretty amazing that WebView  worked without refreshing using JavaScript scrolling.

tonytamsf commented:
> Hmm, I'm not really sure about using a WebView for this. I think we rather want to use TextViews in a RecyclerView. Refreshing a WebView every second will probably cause us trouble later (eg when users scroll while it is refreshing).

I'm trying to understand how I would implement the transcript using RecyclerView + TextView.  

* RecyclerView + LinearLayout with one Textviews for each word?  - I'll have to figure out when and how to break to the next line.  And if the transcription has a long sentence, I'll need to break a sentence into multiple Textviews
* RecyclerView + LinearLayout with one TextView per row? - There seems to be a lot of pixel calculation to figure out the end of a sentence

I would look into using this library for an overflow layout https://github.com/ApmeM/android-flowlayout and count on it to layout the TextViews.  Might be a problem if there is a snippet of really long transcription text than spans more than 1 line can fit.

ByteHamster commented:
> I'm trying to understand how I would implement the transcript using RecyclerView + TextView.

I don't really know either, to be honest. Let's discuss the design in https://github.com/AntennaPod/AntennaPod/issues/4935 and then talk about the technical details here if we know what it should look like

tonytamsf commented:
Closing this PR to start targeting the `transcript` branch
